sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -353,7 +370,58 @@ struct graphs_support;
 } // namespace device
 } // namespace info
 
-class node {};
+enum class dynamic_parameter_type {
+  usm,
+  buffer,
+  scalar
+};
+
+template <dynamic_parameter_type DynParamType>
+class dynamic_parameter {
+public:
+  dynamic_parameter(std::string label = "");
+
+  const std::string& get_label() const;
+};
+
+template <>
+class dynamic_parameter<dynamic_parameter_type::scalar> {
+public:
+  dynamic_parameter(size_t size, std::string label = "");
+
+  size_t get_size() const;
+  const std::string& get_label() const;
+};
+
+class node {
+public:
+  node() = delete;
+
+  /* -- parameter registration interface -- */
+
+  void register_dynamic_parameter(void* param,
```

Comment:
> can we also add register_dynamic_parameter(void* param...) into class command_graph<graph_state::executable>? So we can update the USM memory pointer in a graph created with recording API.

You could also use the `get_node_from_event()` method defined in this PR to get a node object from the event returned in the recording API, and use that with the existing registration APIs.


If you are using the recording API, also how suitable would be using the update API from https://github.com/reble/llvm/pull/343 instead? The plan is to introduce both update APIs.

Reply:
> If you are using the recording API, also how suitable would be using the update API from https://github.com/reble/llvm/pull/343 instead?

I'm afraid we could not use `update(const command_graph<graph_state::modifiable>& graph)`, the new `graph` requires a new recording, if my understanding is correct. The expectation is that we can change the memory pointer in the recorded graph directly, instead of recording again.

> You could also use the get_node_from_event() method defined in this PR to get a node object from the event returned in the recording API

Usually, the reason that we use the recording API is to make the existing codes unchanged (just wrap the existing code with a recording begin/end pair). In pytorch, we could hardly get the event from the existing code.

This answer also implies that there are nodes within the graph even it is recorded, so, we may implement `graph.register_dynamic_parameter` as below:
```
class graph<graph_state::executable> {
     bool register_dynamic_parameter(void *usm, dynamicParam){
         for (node in this->nodes) {
              // ignore if usm is not part of the node.
             node.register_dynamic_parameter(usm, dynamicParam);  
         }
         // return error if none of the node contains the usm.
     }
}
```

Reply:
The concern I have with registering on an executable graph is that by that point the backend representation is already created, so if we allow the user to be able to register/update any node after the backend command-buffer is created, then the backend will have to do more work (some of it unnecessary if we don't update every command parameter) to make that possible.

Would a similar API to the one you've described there, but for a modifiable graph be suitable? e.g `command_graph<modifiable>::register_dynamic_parameter(void *usm, dynamicParam)`. This would allow checking all the nodes without needing the application to manage those node/event objects, but avoid the concern I have with extra runtime overhead by doing this after the backend command-buffer has been created.

Reply:
good idea, we can use the below code to achieve it.
```
command_graph<modifiable>::register_dynamic_parameter(void *usm, dynamicParam)

command_graph<executable>::update_dynamic_parameter(void *newValue1, dynamicParam)
command_graph<executable>::update_dynamic_parameter(void *newValue2, dynamicParam)
```

Reply:
I've moved the register & update APIs to be methods on a `dynamic_parameter` object, and this update method for operate on all nodes in the graph now exists

Reply:
nice change, it is an excellent design to put these APIs in the `class dynamic_parameter`.